src/zap/zaprelocs.cpp

```diff
@@ -120,6 +120,24 @@ void ZapBaseRelocs::WriteReloc(PVOID pSrc, int offset, ZapNode * pTarget, int ta
             PutArm64Rel28((UINT32 *)pLocation,relOffset);
         }
         return;
+
+    case IMAGE_REL_ARM64_PAGEBASE_REL21:
+        {
+            TADDR pSitePage = ((TADDR)m_pImage->GetBaseAddress() + rva) & 0xFFFFFFFFFFFFF000LL;
+            TADDR pActualTargetPage = pActualTarget & 0xFFFFFFFFFFFFF000LL;
+
+            INT64 relPage = (INT64)(pActualTargetPage - pSitePage);
+            INT32 imm21 = (INT32)(relPage >> 12) & 0x1FFFFF;
```

Comment:
Initially I thought the same way, but there is case we use this api to store displacement (target offset)  -- see `recordRelocation` in zapinfo.cpp. So I keep this way.